File: 1035-cousins-in-binary-tree/1035-cousins-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        if root == None:
            return False
        dict = {}
        dict[root.val] = [-1, 0]
        def levelOrder(root, level):
            if root == None:
                return
            if root.left != None:
                dict[root.left.val] = [root.val, level + 1]
            if root.right != None:
                dict[root.right.val] = [root.val, level + 1]
            levelOrder(root.left, level + 1)
            levelOrder(root.right, level + 1)
        levelOrder(root, 1)
        if dict[x][0] != dict[y][0] and dict[x][1] == dict[y][1]:
            return True
        return False

File: 1035-cousins-in-binary-tree/1035-cousins-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<Integer, Pair<Integer, Integer>> map = new HashMap<>();
    public boolean isCousins(TreeNode root, int x, int y) {
        if(root == null) return false;
        map.put(root.val, new Pair(-1, 0));
        levelOrder(root, 0);
        var X = map.get(x);
        var Y = map.get(y);
        return X.getValue() == Y.getValue() && X.getKey() != Y.getKey();
    }
    public void levelOrder(TreeNode root, int level) {
        if(root == null) return;
        if(root.left != null) {
            map.put(root.left.val, new Pair(root.val, level + 1));
        }
        if(root.right != null) {
            map.put(root.right.val, new Pair(root.val, level + 1));
        }
        levelOrder(root.left, level + 1);
        levelOrder(root.right, level + 1);
    }
}